[FLINK-34111][table] Add JSON_QUOTE and JSON_UNQUOTE function

jeyhunkarimov opened the pull request:
## What is the purpose of the change

This pull request adds JSON_QUOTE and JSON_UNQUOTE functions

## Brief change log

  - Add the necessary documentation and functions to support  JSON_QUOTE and JSON_UNQUOTE functions


## Verifying this change

This change added tests and can be verified via `org.apache.flink.table.planner.expressions.ScalarFunctionsTests.testJsonQuote` and `org.apache.flink.table.planner.expressions.ScalarFunctionsTests.testJsonUnquote`


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no 
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no 
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes 
  - If yes, how is the feature documented? docs

jeyhunkarimov commented:
Hi @MartijnVisser could you please review the PR in your available time? Thanks!

xjmdoo commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/SqlFunctionUtils.java:
```diff
@@ -1174,6 +1174,107 @@ public static String toBase64(byte[] bytes) {
+
+        StringBuilder result = new StringBuilder("\"");
+
+        for (int i = 0; i < input.getSizeInBytes(); ++i) {
+            char c = (char) (input.byteAt(i) & 0xFF);
```
I think this doesn't work for multi-byte characters, so it would be better to use code points instead.

jeyhunkarimov replied:
Good point. Fixed

JingGe reviewed (commented):
Thanks @jeyhunkarimov for driving it! I just left some comments.

jeyhunkarimov commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala:
```diff
@@ -604,6 +604,20 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
+  def testJsonQuote(): Unit = {
+    testSqlApi("JSON_QUOTE('null')", "\"null\"")
+    testSqlApi("JSON_QUOTE('\"null\"')", "\"\\\"null\\\"\"")
+    testSqlApi("JSON_QUOTE('[1, 2, 3]')", "\"[1, 2, 3]\"")
```
We can do it. My main intuition was that to escape characters that are reserved in JSON: 
https://www.freeformatter.com/json-escape.html

JingGe replied:
The reason is that https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#json-unquote-character-escape-sequences is referenced in CALCITE-3130

jeyhunkarimov replied:
Makes sense. I updated the `json[Un]quote()` functions and tests accordingly in my last commit.

jeyhunkarimov commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:
```diff
@@ -716,6 +716,28 @@ public SqlSyntax getSyntax() {
                                     SqlTypeFamily.BOOLEAN, SqlTypeFamily.TIME, SqlTypeFamily.TIME)),
                     SqlFunctionCategory.NUMERIC);
 
+    public static final SqlFunction JSON_QUOTE =
```
I have them defined in `BuiltInFunctionDefinitions` (https://github.com/apache/flink/pull/24156/files#diff-95b3db93249837e1786c1af7cd496410d2608bbcf690f64d170145135421ce2eR991) Did I miss something?

JingGe replied:
You are right, I somehow was not aware of that.

JingGe commented:
@flinkbot run azure

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -377,6 +377,12 @@ string:
+    description: Quotes a string as a JSON value by wrapping it with double quote characters, escaping interior quote and other characters, and returning the result as a utf8mb4 string. If the argument is NULL, the function returns NULL.
+  - sql: JSON_UNQUOTE(string)
+    table: STRING.JsonUnquote()
+    description: Unquotes JSON value and returns the result as a utf8mb4 string. If the argument is NULL, returns NULL. If the value starts and ends with double quotes but is not a valid JSON string literal, an error occurs.
```
I didn't get it
I tried ut locally with various of combination and the behaviour doesn't match the description
e.g.
```sql
select JSON_UNQUOTE('"{1,  3]"');
```
returns 
```
{1,  3]
```
did I miss anything?

jeyhunkarimov replied:
Agreed, I borrowed the definition from MySQL. Perhaps we need a json validity check

snuyanzin commented:
I wonder why the implementation doesn't follow new stack implememntation approach like e.g. `org.apache.flink.table.runtime.functions.scalar.CoalesceFunction`, `org.apache.flink.table.runtime.functions.scalar.IfNullFunction` and others inheriting `org.apache.flink.table.runtime.functions.scalar.BuiltInScalarFunction`
?

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/SqlFunctionUtils.java:
```diff
@@ -1174,6 +1174,148 @@ public static String toBase64(byte[] bytes) {
+            int codePoint = inputStr.codePointAt(i);
+            int charCount = Character.charCount(codePoint);
+
+            switch (codePoint) {
```
Escape doesn't seem to follow JSON spec defined at [1]
e.g. it says that `*`, `/` should be escaped and here it doesn't escape it

[1] https://www.json.org/json-en.html

jeyhunkarimov replied:
Here I also used MySQL reference [1]. I am ok to use json.org as a reference

[1] https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#json-unquote-character-escape-sequences

snuyanzin reviewed (changes requested):
Thanks for your contribution 

there is a number of cases where it behaves strange, e.g.
```sql
SELECT  JSON_UNQUOTE('"""');
```

```sql
SELECT  JSON_UNQUOTE('"\"hg"');
```
and other similar


Also if we have a look at json doc[1]
There is a section about escaping

> escape
    '"'
    '\'
    '/'
    'b'
    'f'
    'n'
    'r'
    't'
    'u' hex hex hex hex
hex
    digit
    'A' . 'F'

which seems not being followed by this implementation and not covered by tests.


Also 
I wonder why the implementation doesn't follow new stack implementation approach like e.g. `org.apache.flink.table.runtime.functions.scalar.CoalesceFunction`, `org.apache.flink.table.runtime.functions.scalar.IfNullFunction` and others inheriting `org.apache.flink.table.runtime.functions.scalar.BuiltInScalarFunction`
?


Another issue if input is similar to symbol code but not really a symbol code, for instance
```sql
SELECT  JSON_UNQUOTE('"\u006z"');
```
the current behavior which returns `\u006z006z` here is also wrong

[1] https://www.json.org/json-en.html

jeyhunkarimov commented:
Thanks @snuyanzin  for the review. I will check the tests and add more [failing] tests.About the implementation doesn't follow new stack implementation, I havent noticed that my implementation was using deprecated interfaces/stack. So, there is no specific reason to follow this implementation stack.

snuyanzin commented:
Yes, it's not marked as deprecated 
however for new functions it's better to follow new stack which  is a result of FLIP-32 and FLIP-65.

snuyanzin commented:
Thanks for moving to a new stack

I feel like a lack of IT tests, I would suggest to follow similar approach for tests
Here it is a class for JSON function relates IT tests `org.apache.flink.table.planner.functions.JsonFunctionsITCase`
The reason I asked about that is that I fetched curent branch locally, build it, I can confirm that existing tests are passing, however from user point of `JSON_UNQOTE` does not seem to be working: if you run it e.g. under `flink_sql_client` and execute any query e.g. `SELECT JSON_UNQUOTE('"abc"');` it returns empty string.

Another issue is that current tests are checking only sql (as mentioned not at IT level) and do not check table api

snuyanzin commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -1090,6 +1092,21 @@ public OutType regexpExtract(InType regex) {
+     * Returns a string by quotes a string as a JSON value and wrapping it with double quote
+     * characters.
+     */
+    public OutType jsonQuote(InType input) {
+        return toApiSpecificExpression(
+                unresolvedCall(JSON_QUOTE, toExpr(), objectToExpression(input)));
+    }
+
+    /** Returns utf8mb4 string by unquoting JSON value. */
+    public OutType jsonUnquote(InType input) {
+        return toApiSpecificExpression(
+                unresolvedCall(JSON_UNQUOTE, toExpr(), objectToExpression(input)));
+    }
+
```
How it is supposed to be called from table api?
can you provide a couple of examples?

jeyhunkarimov replied:
Good point. IT cases resolved some issues related to the table API and enabled examples

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -377,6 +377,12 @@ string:
+    table: STRING.JsonQuote()
+    description: Quotes a string as a JSON value by wrapping it with double quote characters, escaping interior quote and other characters, and returning the result as a utf8mb4 string. If the argument is NULL, the function returns NULL.
+  - sql: JSON_UNQUOTE(string)
+    table: STRING.JsonUnquote()
+    description: Unquotes JSON value and returns the result as a utf8mb4 string. If the argument is NULL, returns NULL. If the value starts and ends with double quotes but is not a valid JSON string literal, this value skipped as NULL.
```
implementation still doesn't match documentation for corner cases

also it would make sense to justify this or another behavior by comparing with other vendors

jeyhunkarimov replied:
MariaDB and MySQL have the same definition, the former has more precise about non valid JSON values [1]. 
I adopted that one. 

[1] https://mariadb.com/kb/en/json_unquote/

jeyhunkarimov commented:
> Thanks for moving to a new stack
> 
> I feel like a lack of IT tests, I would suggest to follow similar approach for tests Here it is a class for JSON function relates IT tests `org.apache.flink.table.planner.functions.JsonFunctionsITCase` The reason I asked about that is that I fetched curent branch locally, build it, I can confirm that existing tests are passing, however from user point of `JSON_UNQOTE` does not seem to be working: if you run it e.g. under `flink_sql_client` and execute any query e.g. `SELECT JSON_UNQUOTE('"abc"');` it returns empty string.
> 
> Another issue is that current tests are checking only sql (as mentioned not at IT level) and do not check table api

Thanks for the review. I added IT cases. 

>  `SELECT JSON_UNQUOTE('"abc"');` it returns empty string.
Yes, it returns empty string (now I changed it, based on MySQL/MariaDB to return the same value) because '"abc"' is not a valid json string (according to the definition). For example, '["abc"]' is a valid json string.

MartijnVisser commented:
@snuyanzin Do you want to take another pass?

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -377,6 +377,12 @@ string:
+    description: Quotes a string as a JSON value by wrapping it with double quote characters, escaping interior quote and other characters, and returning the result as a utf8mb4 string. If the argument is NULL, the function returns NULL.
+  - sql: JSON_UNQUOTE(string)
+    table: STRING.JsonUnquote()
+    description: Unquotes JSON value and returns the result as a utf8mb4 string. If the argument is NULL, returns NULL. If the value starts and ends with double quotes but is not a valid JSON string literal, the value is passed through unmodified.
```
IMHO That's looks like a questionable behaviour I would suggest to discuss it first in  jira.
It could be quite obvious to see whether it ws unquoted or or just passed through as unmodified however it could be very tricky for complex strings

Thoughts out loud: By the was we already have `IS JSON`, couldn't we reuse it

jeyhunkarimov replied:
Yes, I also had a similar question. From the definition, if the input is not a valid JSON string literal, then we pass the value as it is. For example, `"key"` is not a valid JSON literal but `["key"]` is. Maybe we need more investigation on this.

snuyanzin commented on flink-python/docs/reference/pyflink.table/expressions.rst:
```diff
@@ -195,6 +195,8 @@ string functions
     Expression.reverse
     Expression.split_index
     Expression.str_to_map
+    Expression.json_quote
+    Expression.json_unquote
```
Shouldn't we put these functions here and in all other python related files together with other json functions?

snuyanzin commented on flink-python/pyflink/table/expression.py:
```diff
@@ -1201,6 +1203,22 @@ def regexp_extract(self,
             return _ternary_op("regexpExtract")(self, regex, extract_index)
 
+    def json_quote(self) -> 'Expression':
+        """
+        Quotes a string as a JSON value by wrapping it with double quote characters,
+        escaping interior quote and other characters, and returning the result as
+        a utf8mb4 string. If the argument is NULL, the function returns NULL.
+        """
+        return _unary_op("jsonQuote")(self)
+
+    def json_unquote(self) -> 'Expression':
+        """
+        Unquotes JSON value and returns the result as a utf8mb4 string.
+        If the argument is NULL, returns NULL. If the value starts and ends with
+        double quotes but is not a valid JSON string literal, an error occurs.
+        """
+        return _unary_op("jsonUnquote")(self)
+
```
Same issue as mentioned for documentation

snuyanzin reviewed (changes requested):
Based on current description if I understand it correctly (please correct me if I'm wrong): I assume that if my input is correct and quoted then I would also assume that after applying of `JSON_UNQUOTE` I can receive the initial string

However
```sql
SELECT json_unquote(json_quote('"key": "value"'));
```
returns  `"\"key\": \"value\""`
I think this should be fixed to be able to get again `'"key": "value"'` ?
Or did I miss anything?

jeyhunkarimov commented:
Thanks for the comment @snuyanzin . I think we need more investigation about the clear definition and behavior of these functions, since the ones adopted from MySQL are a bit ambiguous. 

> Based on current description if I understand it correctly (please correct me if I'm wrong): I assume that if my input is correct and quoted then I would also assume that after applying of `JSON_UNQUOTE` I can receive the initial string
> 
> However
> 
> ```sql
> SELECT json_unquote(json_quote('"key": "value"'));
> ```
> 
> returns `"\"key\": \"value\""` I think this should be fixed to be able to get again `'"key": "value"'` ? Or did I miss anything?

Is it mainly because of the definition of `JSON_QUOTE` that does not require its input to be valid JSON, and `JSON_UNQOTE` that require its input to be a valid JSON. 

As I mentioned, we need a clearer definition (maybe not from MySQL references)

snuyanzin commented:
>Is it mainly because of the definition of JSON_QUOTE that does not require its input to be valid JSON, and JSON_UNQOTE that require its input to be a valid JSON.

I'm not sure that this is the only reson

there is a more simple case, for instance '"key"` is a valid JSON and `json_quote` produces also a valid json `"\"key\""`
however the combination also seems not working
```sql
SELECT json_unquote(json_quote('"key"'));
```

jeyhunkarimov commented:
> > Is it mainly because of the definition of JSON_QUOTE that does not require its input to be valid JSON, and JSON_UNQOTE that require its input to be a valid JSON.
> 
> I'm not sure that this is the only reason
> 
> there is a more simple case, for instance `"key"` is a valid JSON and `json_quote` produces also a valid json `"\"key\""` however the combination also seems not working
> 
> ```sql
> SELECT json_unquote(json_quote('"key"'));
> ```
> 
> and continues producing `"\"key\""`

Let me paste the definitions here:

`json_quote`: Quotes a string as a JSON value by wrapping it with double quote characters, escaping interior quote and other characters, and returning the result as a utf8mb4 string. If the argument is NULL, the function returns NULL.

j`son_unquote`: Unquotes JSON value and returns the result as a utf8mb4 string. If the argument is NULL, returns NULL. If the value starts and ends with double quotes but is not a valid JSON string literal, the value is passed through unmodified.

So, from this definition, `json_quote` does not enforce the input to be a valid json, but `json_unquote` does. 

So, in your example, `json_quote('"key"')`  produces `"\"key\""` and `json_unqote("\"key\"")` will produce the same `("\"key\"")` since the input `("\"key\"")` is not a valid json. 

Instead this works: `testSqlApi("JSON_UNQUOTE(JSON_QUOTE('[1,2,3]'))", "[1,2,3]")` since the input of the json_unqote is a valid json.

These rules I derived from the definition I pasted above. Maybe I am missing something.

snuyanzin commented:
> and json_unqote("\\"key\\"") will produce the same ("\\"key\\"") since the input ("\\"key\\"") is not a valid json.

can you clarify why it the input here is not a valid json?

Based on [1]
valid  json is an element
> json
>        element

which is a value separated by whitespaces
> element
    ws value ws

and value itself could be a string    
>  value
   object
   array
   string
   number
   "true"
   "false"
   "null"

which is characters
>string
'"' characters '"'

>characters
    ""
    character characters
and characters could contain escape characters

>character
    '0020' . '10FFFF' - '"' - '\'
'\' escape

like that 
>escape
    '"'
    '\'
    '/'
    'b'
    'f'
    'n'
    'r'
    't'
    'u' hex hex hex hex

so I didn't get why  `"\"key\""` should be considered as an invalid json...
Can you elaborate here please?

By the way I checked these queries against MySQL and it works as expected [2]
```SQL
select json_unquote(json_quote("\"key\""));
select json_unquote(json_quote("\"key\": \"value\""));
```
it returns
```
"key"
"key": "value"
```

[1] https://www.json.org/json-en.html
[2] https://www.db-fiddle.com/f/fuuT4xzfFEbY772eNkLkmp/0

jeyhunkarimov commented:
Thanks a lot for the comments. I agree. I was a bit influenced by the definition of MySQL and didn't check the `json.org`.

snuyanzin reviewed (changes requested):
I checked that
```sql
select json_unquote(json_quote('"key"'));
```
now returns `"key"` which is expected

however
```sql
select json_quote('"key"');
```
returns `""key""` which is NOT OK since nested quotes are not escaped
I would expect something like
```sql
"\"key\""
```
and MySQL returns this

similar issues with other combinations of symbols which should be escapoed

anupamaggarwal commented:
Hi @jeyhunkarimov thank you for contributing this PR!. Just wanted to check if you were planning on working further on this,  otherwise I'd be happy to continue and address some of the feedback. Please let me know 👍

jeyhunkarimov commented:
Hi @anupamaggarwal thanks for your proposal. Please feel free to continue!

anupamaggarwal commented:
thanks @jeyhunkarimov! linking the PR https://github.com/apache/flink/pull/24967

snuyanzin commented:
superseded by https://github.com/apache/flink/pull/24967
